roxie/udplib/udptrs.cpp

```diff
@@ -34,73 +36,156 @@
 // MORE - why use DataBuffers on output side?? We could use zeroCopy techniques if we had a dedicated memory area.
-
-class UdpReceiverEntry 
+// But using them on this side means we guarantee that the packets fit into databuffers on the other side... But so would matching their size
+
+/*
+ *
+ * There are 3 threads running to manage the data transfer from slave back to server:
+ * send_resend_flow
+ *   - checks periodically that nothing is waiting for a "request to send" that timed out
+ * send_receive_flow
+ *   - waits on socket receiving "ok_to_send" packets from servers
+ *   - updates state of relevant receivers
+ *   - pushes permission tokens to a queue
+ * send_data
+ *   - waits on queue of permission tokens
+ *   - broadcasts "busy"
+ *   - writes data to server
+ *   - broadcasts "no longer "
+ *   - sends "completed" or "completed but I want to send more" flow message to server
+ *
+ * Queueing up data packets is done by the slave worker threads.
+ * *
+
+ *
+ * Data races to watch for
+ * 1. Two slave threads add data at same time - only one should sent rts (use atomic_inc for the count)
+ * 2. We check for timeout and resend rts or fail just as permission comes in
+ *    - resend rts is harmless ?
+ *    - fail is acceptable
+ * 3. After sending data, we need to decide whether to set state to 'pending' (and send rts) or empty. If we read count, decide it's zero
+ *    and then (before we set state) someone adds data (and sends rts), we must not set state to empty. CAS to set state empty only if
+ *    it's sending_data perhaps?
+ * 4. While sending data, someone adds new data. They need to send rts and set state to pending whether empty or sending_data
+ * 5. Do we need sending_data state? Is it the same as empty, really? Is empty the same as 'count==0' ? Do we even need state?
+ *    - send rts whenever incrementing count from zero
+ *    - resend rts if count is non-zero and timed out
+ *    - resend rts if we send data but there is some remaining
+ */
+
+class UdpReceiverEntry : public IUdpReceiverEntry
 {
-    queue_t *output_queue;
-    bool    initialized;
-
-public:
-    ISocket *send_flow_socket;
-    ISocket *data_socket;
-    unsigned numQueues;
-    int     current_q;
-    int     currentQNumPkts;   // Current Queue Number of Consecutive Processed Packets.
-    int     *maxPktsPerQ;      // to minimise power function re-calc for evey packet
-
-    // MORE - consider where we need critsecs in here!
-
-    void sendRequest(unsigned myNodeIndex, flow_t::flowmsg_t cmd)
-    {
-        UdpRequestToSendMsg msg = {sizeof(UdpRequestToSendMsg), static_cast<unsigned short>(cmd), static_cast<unsigned short>(myNodeIndex), 0};
-        try 
-        {
-            send_flow_socket->write(&msg, msg.length);
+private:
+    queue_t *output_queue = nullptr;
+    bool    initialized = false;
+    const bool isLocal = false;
+    ISocket *send_flow_socket = nullptr;
+    ISocket *data_socket = nullptr;
+    const unsigned numQueues = 0;
```

Comment:
I think you do get an error if a const member is not initialized

Reply:
Confirmed. I don't have especially strong feelings either way. In this case it would probably be preferable to NOT initialize here so that we see an error if not set in a constructor.

Reply:
You could argue the same for non-const members - should you initialize to an invalid value, or should you turn on the compiler option to give an error if not set in the constructor?